File: src/lib/pet-memory.ts
"use client";

/**
 * 跨 session 宠物记忆 —— Claudio 不再每次启动都失忆。
 *
 * 设计原则(很重要,以后改的人请读):
 *
 *   ❌ **不要**把"偏爱艺人 / 跳过率高艺人"喂给 AI prompt。
 *      用户的库本身就是 TA 已经亲自挑选的 —— 库里的歌默认都是 TA 喜欢的。
 *      把 skip 解释成"不喜欢"会把"瞬时心情不对"误读成"长期排斥",
 *      AI 会越来越避开那个艺人 → TA 再也听不到 → 库丰富度被悄悄削平。
 *      反向同理: 把多次完整播完当"偏爱"会让 AI 反复推荐,造成单调。
 *      这是推荐系统经典的死亡螺旋。
 *
 *   ✅ **可以**用 behavior 数据做"近期已播过 → 暂时降权"(变化,不是偏好)。
 *      这种已经在 candidate-ranker 里做了,用于增加多样性,**不是**让 AI 学口味。
 *
 *   ✅ **可以**记累积事实: 总播放数、首次相识时间、用户最近说过的话。
 *      这些是"陪伴的痕迹",不是把 TA 框成某种类型。
 *
 * 数据来源:
 *   1. 持久化的"最近 N 条用户说过的话"(去掉招呼/感谢) —— 让它能说"上次你提过…"
 *   2. 累积统计 —— 总播放数 / 首次相识日期
 *
 * 输出: 一句话 digest, 注入到 AI 三个调用点的 USER prompt 里(SYSTEM 不能动,会破坏缓存)。
 */

import { cache } from "./tauri";
import { readBehaviorLog } from "./behavior-log";

const MEM_KEY = "pet_memory_v1";
const MAX_UTTERANCES = 30; // 最多记 30 条用户原话, 老的丢
const UTTERANCE_TTL_DAYS = 30; // 30 天没更新的彻底过期

type PersistedMemory = {
  version: 1;
  /** 用户跟 Claudio 说过的有意义的话(过滤掉招呼/感谢/单字) */
  utterances: { ts: number; text: string }[];
  /** 第一次见到 TA 的时间(unix sec) */
  firstSeenAt: number;
  /** 最后一次见到的时间 */
  lastSeenAt: number;
  /**
   * 用户自述的关于自己的情况:工作时间、作息、习惯、爱好...
   * 用户在设置里手动填的,不是 AI 推断的。**TA 自己说的**是 ground truth,
   * 应该比从行为日志反推的"偏好"权重高得多。
   */
  userFacts?: string;
};

let memo: PersistedMemory | null = null;

async function loadMemory(): Promise<PersistedMemory> {
  if (memo) return memo;
  try {
    const raw = await cache.getState(MEM_KEY);
    if (raw) {
      const parsed = JSON.parse(raw) as PersistedMemory;
      if (parsed.version === 1) {
        memo = parsed;
        return memo;
      }
    }
  } catch {
    /* 坏缓存 → 重置 */
  }
  const now = Math.floor(Date.now() / 1000);
  memo = {
    version: 1,
    utterances: [],
    firstSeenAt: now,
    lastSeenAt: now,
  };
  return memo;
}

async function saveMemory(): Promise<void> {
  if (!memo) return;
  try {
    await cache.setState(MEM_KEY, JSON.stringify(memo));
  } catch (e) {
    console.debug("[claudio] pet-memory 落盘失败", e);
  }
}

/**
 * 用户每次说话后调一次。**自动过滤**掉无意义短句(招呼/感谢/单字),不污染记忆。
 */
export async function recordUserUtterance(text: string): Promise<void> {
  const trimmed = text.trim();
  if (!trimmed) return;
  if (trimmed.length <= 2) return; // 单字短句没价值
  if (/^(嗯|哦|好|谢谢|你好|您好|早|晚安|hi|hello|thanks|thx)$/i.test(trimmed)) return;

  const m = await loadMemory();
  const now = Math.floor(Date.now() / 1000);
  m.utterances.push({ ts: now, text: trimmed.slice(0, 80) });
  // 滚动 cap + 过期清理
  const cutoff = now - UTTERANCE_TTL_DAYS * 86400;
  m.utterances = m.utterances.filter((u) => u.ts >= cutoff).slice(-MAX_UTTERANCES);
  m.lastSeenAt = now;
  await saveMemory();
}

/**
 * 给 AI prompt 用的一行 digest。
 * 包含:陪伴痕迹(总播放数 / 上次说过的话) + 用户自述事实(若有)。
 *
 * 例: "听过 156 首 · 上次说'今天好累'(2 小时前) · 自述:工作日 9-18:30,双休"
 *
 * 故意**不**包含 loveArtists / skipHotArtists —— 见模块顶部注释。
 */
export async function getMemoryDigest(): Promise<string> {
  const [m, events] = await Promise.all([loadMemory(), readBehaviorLog()]);
  const parts: string[] = [];

  if (events.length > 0) {
    parts.push(`听过 ${events.length} 首`);
  }

  // 最近一句有意义的话 —— 这是真正能让 Claudio 显得"记得"的信号
  const lastUtt = m.utterances[m.utterances.length - 1];
  if (lastUtt) {
    const ageMin = Math.floor((Date.now() / 1000 - lastUtt.ts) / 60);
    let ageLabel: string;
    if (ageMin < 5) ageLabel = "刚刚";
    else if (ageMin < 60) ageLabel = `${ageMin} 分钟前`;
    else if (ageMin < 24 * 60) ageLabel = `${Math.floor(ageMin / 60)} 小时前`;
    else ageLabel = `${Math.floor(ageMin / (60 * 24))} 天前`;
    parts.push(`上次说"${lastUtt.text.slice(0, 30)}"(${ageLabel})`);
  }

  // 用户自述事实——TA 自己说的,比任何推断都准
  if (m.userFacts && m.userFacts.trim()) {
    parts.push(`自述:${m.userFacts.trim()}`);
  }

  return parts.join(" · ");
}

// ---- 用户自述 (settings 页编辑) ----

export async function getUserFacts(): Promise<string> {
  const m = await loadMemory();
  return m.userFacts ?? "";
}

export async function setUserFacts(text: string): Promise<void> {
  const m = await loadMemory();
  // 留点空间但不要无限大 —— 400 字够说工作时间、作息、爱好、几个小习惯
  m.userFacts = text.trim().slice(0, 400);
  await saveMemory();
}

/** 调试 / 设置页用 —— 拿到完整记忆。 */
export async function readPetMemory(): Promise<PersistedMemory> {
  return loadMemory();
}

/** 用户主动清记忆(以后设置页可以加按钮)。 */
export async function clearPetMemory(): Promise<void> {
  const now = Math.floor(Date.now() / 1000);
  memo = {
    version: 1,
    utterances: [],
    firstSeenAt: now,
    lastSeenAt: now,
  };
  await saveMemory();
}

File: android-native/app/src/main/java/app/pipo/nativeapp/data/PetMemory.kt
package app.pipo.nativeapp.data

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

/**
 * 跨 session 宠物记忆 —— 镜像 src/lib/pet-memory.ts。
 *
 * 设计原则（和 React 版完全一致，请别加"喂偏好艺人给 AI"这种逻辑——
 * 那是死亡螺旋开端，TS 顶部注释里专门写了原因）：
 *
 *   ❌ 不喂"偏爱/跳过率高"艺人给 AI：把瞬时心情误读成长期排斥，库丰富度被悄悄削平
 *   ✅ 记累积事实 + 最近原话：陪伴痕迹，不是把 TA 框成某种类型
 *   ✅ userFacts：TA 自述比任何推断都准
 */
class PetMemory(context: Context) {
    private val prefs = context.applicationContext
        .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    /** 后台 IO 调度器 —— SharedPrefs JSON 构建 + 写盘都不能阻塞主线程 */
    private val ioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    data class Memory(
        val version: Int,
        val utterances: MutableList<Utterance>,
        val conversation: MutableList<ConversationTurn>,
        val musicReferences: MutableList<MusicReference>,
        var firstSeenAt: Long,
        var lastSeenAt: Long,
        var userFacts: String,
        var conversationSummary: String,
    )

    data class Utterance(val tsSec: Long, val text: String)
    data class ConversationTurn(val role: String, val text: String, val tsSec: Long)
    data class MusicReference(
        val title: String,
        val artist: String = "",
        val reason: String = "",
        val tsSec: Long = 0L,
    )
    data class ConversationContext(
        val summary: String = "",
        val turns: List<ConversationTurn> = emptyList(),
        val musicReferences: List<MusicReference> = emptyList(),
    )

    @Volatile
    private var memo: Memory? = null

    @Synchronized
    private fun load(): Memory {
        memo?.let { return it }
        val raw = prefs.getString(KEY, null)
        val parsed = raw?.let { runCatching { JSONObject(it) }.getOrNull() }
        if (parsed != null && parsed.optInt("version") == VERSION) {
            val utt = mutableListOf<Utterance>()
            val arr = parsed.optJSONArray("utterances")
            if (arr != null) {
                for (i in 0 until arr.length()) {
                    val o = arr.optJSONObject(i) ?: continue
                    utt.add(Utterance(o.optLong("ts"), o.optString("text")))
                }
            }
            val conv = mutableListOf<ConversationTurn>()
            val convArr = parsed.optJSONArray("conversation")
            if (convArr != null) {
                for (i in 0 until convArr.length()) {
                    val o = convArr.optJSONObject(i) ?: continue
                    val role = normalizeRole(o.optString("role"))
                    val text = cleanConversationText(o.optString("text"))
                    if (role != null && text.isNotBlank()) {
                        conv.add(ConversationTurn(role, text, o.optLong("ts")))
                    }
                }
            }
            val refs = mutableListOf<MusicReference>()
            val refArr = parsed.optJSONArray("musicReferences")
            if (refArr != null) {
                for (i in 0 until refArr.length()) {
                    val o = refArr.optJSONObject(i) ?: continue
                    cleanMusicReference(
                        title = o.optString("title"),
                        artist = o.optString("artist"),
                        reason = o.optString("reason"),
                        tsSec = o.optLong("ts"),
                    )?.let { refs.add(it) }
                }
            }
            memo = Memory(
                version = VERSION,
                utterances = utt,
                conversation = conv.takeLast(MAX_RAW_CONVERSATION_TURNS).toMutableList(),
                musicReferences = refs.takeLast(MAX_MUSIC_REFERENCES).toMutableList(),
                firstSeenAt = parsed.optLong("firstSeenAt"),
                lastSeenAt = parsed.optLong("lastSeenAt"),
                userFacts = parsed.optString("userFacts"),
                conversationSummary = parsed.optString("conversationSummary").take(MAX_SUMMARY_CHARS),
            )
        } else {
            val now = System.currentTimeMillis() / 1000
            memo = Memory(
                version = VERSION,
                utterances = mutableListOf(),
                conversation = mutableListOf(),
                musicReferences = mutableListOf(),
                firstSeenAt = now,
                lastSeenAt = now,
                userFacts = "",
                conversationSummary = "",
            )
        }
        return memo!!
    }

    @Synchronized
    private fun save() {
        val m = memo ?: return
        val arr = JSONArray()
        m.utterances.forEach {
            arr.put(JSONObject().apply { put("ts", it.tsSec); put("text", it.text) })
        }
        val convArr = JSONArray()
        m.conversation.forEach {
            convArr.put(JSONObject().apply {
                put("role", it.role)
                put("text", it.text)
                put("ts", it.tsSec)
            })
        }
        val refArr = JSONArray()
        m.musicReferences.forEach {
            refArr.put(JSONObject().apply {
                put("title", it.title)
                put("artist", it.artist)
                put("reason", it.reason)
                put("ts", it.tsSec)
            })
        }
        val obj = JSONObject().apply {
            put("version", m.version)
            put("utterances", arr)
            put("conversation", convArr)
            put("musicReferences", refArr)
            put("firstSeenAt", m.firstSeenAt)
            put("lastSeenAt", m.lastSeenAt)
            put("userFacts", m.userFacts)
            put("conversationSummary", m.conversationSummary)
        }
        prefs.edit().putString(KEY, obj.toString()).apply()
    }

    /**
     * 用户每次说话调一次。fire-and-forget —— JSON 构建 + 写盘都丢到 IO 调度器，
     * 主线程从 onSend 调用时不会被 SharedPrefs 写阻塞（之前在 30 条 utterance 时能 ~50ms）。
     * 自动过滤无意义短句。
     */
    fun recordUtterance(text: String) {
        val trimmed = text.trim()
        if (trimmed.length <= 2) return
        if (Regex("^(嗯|哦|好|谢谢|你好|您好|早|晚安|hi|hello|thanks|thx)$", RegexOption.IGNORE_CASE).matches(trimmed)) return

        ioScope.launch {
            val m = load()
            val now = System.currentTimeMillis() / 1000
            m.utterances.add(Utterance(now, trimmed.take(80)))
            val cutoff = now - UTTERANCE_TTL_DAYS * 86400L
            val filtered = m.utterances.filter { it.tsSec >= cutoff }.takeLast(MAX_UTTERANCES)
            m.utterances.clear()
            m.utterances.addAll(filtered)
            m.lastSeenAt = now
            save()
        }
    }

    /** 最后一句有意义的话 + 上次时间标签 */
    fun lastUtterance(): Utterance? = load().utterances.lastOrNull()

    fun conversationContext(): ConversationContext {
        val m = load()
        return ConversationContext(
            summary = m.conversationSummary,
            turns = m.conversation.toList(),
            musicReferences = m.musicReferences.toList(),
        )
    }

    suspend fun recordConversationTurn(role: String, text: String) {
        withContext(Dispatchers.IO) {
            recordConversationTurnBlocking(role, text)
        }
    }

    @Synchronized
    private fun recordConversationTurnBlocking(role: String, text: String) {
        val normalizedRole = normalizeRole(role) ?: return
        val cleaned = cleanConversationText(text)
        if (cleaned.isBlank()) return
        val m = load()
        val now = System.currentTimeMillis() / 1000
        m.conversation.add(ConversationTurn(normalizedRole, cleaned, now))
        trimConversation(m)
        m.lastSeenAt = now
        save()
    }

    suspend fun recordMusicReferences(references: List<MusicReference>) {
        if (references.isEmpty()) return
        withContext(Dispatchers.IO) {
            recordMusicReferencesBlocking(references)
        }
    }

    @Synchronized
    private fun recordMusicReferencesBlocking(references: List<MusicReference>) {
        val cleaned = references.mapNotNull {
            cleanMusicReference(it.title, it.artist, it.reason, it.tsSec)
        }
        if (cleaned.isEmpty()) return
        val m = load()
        val now = System.currentTimeMillis() / 1000
        val byKey = LinkedHashMap<String, MusicReference>()
        for (old in m.musicReferences) {
            val key = musicReferenceKey(old)
            if (key.isNotBlank()) byKey[key] = old
        }
        for (ref in cleaned) {
            val stamped = ref.copy(tsSec = if (ref.tsSec > 0L) ref.tsSec else now)
            val key = musicReferenceKey(stamped)
            if (key.isBlank()) continue
            byKey.remove(key)
            byKey[key] = stamped
        }
        val cutoff = now - MUSIC_REFERENCE_TTL_DAYS * 86400L
        val fresh = byKey.values
            .filter { it.tsSec >= cutoff }
            .takeLast(MAX_MUSIC_REFERENCES)
        m.musicReferences.clear()
        m.musicReferences.addAll(fresh)
        m.lastSeenAt = now
        save()
    }

    fun firstSeenAt(): Long = load().firstSeenAt

    fun userFacts(): String = load().userFacts

    fun setUserFacts(text: String) {
        val m = load()
        m.userFacts = text.trim().take(400)
        save()
    }

    fun clear() {
        val now = System.currentTimeMillis() / 1000
        memo = Memory(VERSION, mutableListOf(), mutableListOf(), mutableListOf(), now, now, "", "")
        save()
    }

    /**
     * 只清对话相关记忆（对话流 / 摘要 / 最近原话 / 可执行音乐指代），**保留** userFacts
     * （ABOUT YOU 是用户显式编辑的画像，不该被「清空对话」误删）和 firstSeenAt（相识时间）。
     * 供设置页「清空 AI 对话记忆」调用，与 PetChatStore.clear() 一起把界面与底层都归零。
     */
    @Synchronized
    fun clearConversation() {
        val m = load()
        m.conversation.clear()
        m.musicReferences.clear()
        m.utterances.clear()
        m.conversationSummary = ""
        m.lastSeenAt = System.currentTimeMillis() / 1000
        save()
    }

    /**
     * AI prompt 用的一行 digest。
     * 镜像 React getMemoryDigest：陪伴痕迹（总播放数 + 上次说过的话）+ 用户自述
     * 故意**不**包含 loveArtists / skipHotArtists —— 见 React 顶部注释
     */
    fun digest(behaviorTotalEvents: Int, tasteSummary: String? = null): String {
        val m = load()
        val parts = mutableListOf<String>()
        if (behaviorTotalEvents > 0) parts.add("听过 $behaviorTotalEvents 首")
        m.utterances.lastOrNull()?.let { last ->
            val ageMin = ((System.currentTimeMillis() / 1000 - last.tsSec) / 60).toInt()
            val ageLabel = when {
                ageMin < 5 -> "刚刚"
                ageMin < 60 -> "$ageMin 分钟前"
                ageMin < 24 * 60 -> "${ageMin / 60} 小时前"
                else -> "${ageMin / (60 * 24)} 天前"
            }
            parts.add("上次说\"${last.text.take(30)}\"($ageLabel)")
        }
        if (m.userFacts.isNotBlank()) parts.add("自述:${m.userFacts.trim()}")
        if (!tasteSummary.isNullOrBlank()) parts.add(tasteSummary.trim())
        return parts.joinToString(" · ")
    }

    private fun trimConversation(m: Memory) {
        if (m.conversation.size <= MAX_RAW_CONVERSATION_TURNS) return
        val moveCount = (m.conversation.size - RAW_CONVERSATION_KEEP_TURNS)
            .coerceAtLeast(1)
            .coerceAtMost(m.conversation.size)
        val moved = m.conversation.take(moveCount)
        val remaining = m.conversation.drop(moveCount)
        val addition = summarizeConversationTurns(moved)
        m.conversationSummary = compactSummary(m.conversationSummary, addition)
        m.conversation.clear()
        m.conversation.addAll(remaining)
    }

    private fun summarizeConversationTurns(turns: List<ConversationTurn>): String {
        val lines = mutableListOf<String>()
        var pendingUser: String? = null
        for (turn in turns) {
            when (turn.role) {
                ROLE_USER -> pendingUser = turn.text
                ROLE_ASSISTANT -> {
                    val u = pendingUser
                    if (!u.isNullOrBlank()) {
                        lines.add("用户:${u.take(54)} / Pipo:${turn.text.take(54)}")
                        pendingUser = null
                    } else {
                        lines.add("Pipo:${turn.text.take(70)}")
                    }
                }
            }
        }
        if (!pendingUser.isNullOrBlank()) lines.add("用户:${pendingUser.take(70)}")
        return lines.joinToString("\n")
    }

    private fun compactSummary(existing: String, addition: String): String {
        val combined = listOf(existing.trim(), addition.trim())
            .filter { it.isNotBlank() }
            .joinToString("\n")
        if (combined.length <= MAX_SUMMARY_CHARS) return combined
        return combined
            .lines()
            .filter { it.isNotBlank() }
            .takeLast(MAX_SUMMARY_LINES)
            .joinToString("\n")
            .takeLast(MAX_SUMMARY_CHARS)
            .trim()
    }

    private fun cleanConversationText(text: String): String =
        text.replace('\n', ' ')
            .replace(Regex("\\s+"), " ")
            .trim()
            .take(MAX_CONVERSATION_TEXT_CHARS)

    private fun cleanMusicReference(
        title: String,
        artist: String,
        reason: String,
        tsSec: Long,
    ): MusicReference? {
        val t = cleanConversationText(title).take(80)
        if (t.isBlank() || t == "null") return null
        val a = cleanConversationText(artist).take(80).takeUnless { it == "null" }.orEmpty()
        val r = cleanConversationText(reason).take(80).takeUnless { it == "null" }.orEmpty()
        return MusicReference(title = t, artist = a, reason = r, tsSec = tsSec)
    }

    private fun musicReferenceKey(ref: MusicReference): String =
        (ref.title + "|" + ref.artist).lowercase()
            .replace(Regex("[\\s'\"`·・\\-－—_,，。.、!?！？()（）\\[\\]【】《》<>&/]+"), "")

    private fun normalizeRole(role: String): String? =
        when (role.trim().lowercase()) {
            ROLE_USER -> ROLE_USER
            ROLE_ASSISTANT -> ROLE_ASSISTANT
            else -> null
        }

    companion object {
        const val ROLE_USER = "user"
        const val ROLE_ASSISTANT = "assistant"
        private const val PREFS_NAME = "claudio_pet_memory"
        private const val KEY = "v1"
        private const val VERSION = 1
        private const val MAX_UTTERANCES = 30
        private const val UTTERANCE_TTL_DAYS = 30
        private const val MAX_RAW_CONVERSATION_TURNS = 18
        private const val RAW_CONVERSATION_KEEP_TURNS = 12
        private const val MAX_CONVERSATION_TEXT_CHARS = 220
        private const val MAX_SUMMARY_CHARS = 900
        private const val MAX_SUMMARY_LINES = 12
        private const val MAX_MUSIC_REFERENCES = 16
        private const val MUSIC_REFERENCE_TTL_DAYS = 7
    }
}
